### PR TITLE
feat(api): inbox - change all notifications status

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?
Three endpoints to change all notifications statuses, based on [the API DX guide](https://www.notion.so/novuhq/The-new-internal-Inbox-REST-APIs-efa2c9b4592a4f5e9b20011f9202a7c3?pvs=4#1f807efe343b400a9d1644d343bcf4c0):
- update all to read: `POST /inbox/notifications/all-read`
- update all to archived: `POST /inbox/notifications/all-archived`
- update all read to archived: `POST /inbox/notifications/all-read-archived`

### Screenshots

https://github.com/novuhq/novu/assets/2607232/f39e667a-2ec2-4c30-9811-00068241b3ea

https://github.com/novuhq/novu/assets/2607232/0718873c-ac47-423a-912d-0f92ca12ab4e

https://github.com/novuhq/novu/assets/2607232/9900aa27-6cb1-4fc9-a480-ff57792e26df

### CODE REVIEW

File: apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    if (!subscriber) {
+      throw new ApiException(`Subscriber with id: ${command.subscriberId} is not found.`);
+    }
+
+    await this.messageRepository.updateMessagesFromToStatus({
+      environmentId: command.environmentId,
+      subscriberId: subscriber._id,
+      from: command.from,
+      to: command.to,
+    });
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts:43]: the update logic is hidden under the repository, from the use-case level we just say "update notifications with this status to that"

File: apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        _environmentId: command.environmentId,
+      }),
+    });
+
+    await this.invalidateCache.invalidateQuery({
+      key: buildMessageCountKey().invalidate({
+        subscriberId: command.subscriberId,
+        _environmentId: command.environmentId,
+      }),
+    });
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts:57]: after the status is updated we need to invalidate the "feed cache"

File: apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    this.webSocketsQueueService.add({
+      name: 'sendMessage',
+      data: {
+        event: WebSocketEventEnum.UNREAD,
+        userId: subscriber._id,
+        _environmentId: command.environmentId,
+      },
+      groupId: subscriber._organizationId,
+    });
```
Comment (LetItRock) [inline apps/api/src/app/inbox/usecases/update-all-notifications/update-all-notifications.usecase.ts:74]: send the ws event that unread message count has changed

File: libs/dal/src/repositories/message/message.repository.ts [hunk 4]
```diff
@@ -451,17 +531,13 @@ export class MessageRepository extends BaseRepository<MessageDBModel, MessageEnt
    * unread -> { seen: true, read: false, archived: false }
    * unarchived -> { seen: true, read: true, archived: false }
    */
-  async updateMessagesStatus({
-    environmentId,
-    subscriberId,
-    ids,
+  private async updateMessagesStatus({
```
Comment (LetItRock) [inline libs/dal/src/repositories/message/message.repository.ts:534]: I've changed this to be more reusable

File: libs/dal/src/repositories/message/message.repository.ts [hunk 5]
```diff
@@ -451,17 +531,13 @@ export class MessageRepository extends BaseRepository<MessageDBModel, MessageEnt
...
+    query,
     seen,
     read,
     archived,
   }: {
-    environmentId: string;
-    subscriberId: string;
-    ids: string[];
+    query: MessageQuery & EnforceEnvId;
```
Comment (LetItRock) [inline libs/dal/src/repositories/message/message.repository.ts:540]: `MessageQuery & EnforceEnvId` these types are set on the repository, hence the query is using it

File: apps/api/src/app/inbox/dtos/update-all-notifications-request.dto.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { IsArray, IsOptional, IsString } from 'class-validator';
+
+import type { NotificationFilter } from '../utils/types';
+
+export class UpdateAllNotificationsRequestDto implements Pick<NotificationFilter, 'tags'> {
+  @IsOptional()
+  @IsArray()
+  @IsString({ each: true })
+  tags?: string[];
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/dtos/update-all-notifications-request.dto.ts:9]: Do we have allowed characters limitations on tags? If yes, we should enforce the rules on the validation logic and the TS layer with a custom tag type.

File: apps/api/src/app/inbox/inbox.controller.ts [hunk 7]
```diff
@@ -195,4 +199,71 @@ export class InboxController {
       })
     );
   }
+
+  @UseGuards(AuthGuard('subscriberJwt'))
+  @Post('/notifications/all-read')
+  @HttpCode(HttpStatus.NO_CONTENT)
+  async updateAllStatusToRead(
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/inbox.controller.ts]:
```suggestion
  async markAllAsRead(
```

File: apps/api/src/app/inbox/inbox.controller.ts [hunk 8]
```diff
@@ -195,4 +199,71 @@ export class InboxController {
       })
     );
   }
+
+  @UseGuards(AuthGuard('subscriberJwt'))
+  @Post('/notifications/all-read')
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/inbox.controller.ts]:
```suggestion
  @Post('/notifications/mark-all-as-read')
```

If not mistaken, that's where we landed. the DX guide should be updated accordingly.

File: apps/api/src/app/inbox/inbox.controller.ts [hunk 9]
```diff
@@ -195,4 +199,71 @@ export class InboxController {
...
+        },
+        to: {
+          read: true,
+        },
+      })
+    );
+  }
+
+  @UseGuards(AuthGuard('subscriberJwt'))
+  @Post('/notifications/all-archived')
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/inbox.controller.ts]:
```suggestion
  @Post('/notifications/mark-all-as-archived')
```

File: apps/api/src/app/inbox/inbox.controller.ts [hunk 10]
```diff
@@ -195,4 +199,71 @@ export class InboxController {
...
+        },
+        to: {
+          archived: true,
+        },
+      })
+    );
+  }
+
+  @UseGuards(AuthGuard('subscriberJwt'))
+  @Post('/notifications/all-read-archived')
```
Comment (SokratisVidros) [inline apps/api/src/app/inbox/inbox.controller.ts]:
```suggestion
  @Post('/notifications/mark-all-as-read-archived')
```

**Review by SokratisVidros — APPROVED** [review]

File: apps/api/src/app/inbox/inbox.controller.ts (hunk 8 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/inbox.controller.ts]: It's not a problem, I'll update it :)

File: apps/api/src/app/inbox/dtos/update-all-notifications-request.dto.ts (hunk 6 above)
Comment (LetItRock) [reply apps/api/src/app/inbox/dtos/update-all-notifications-request.dto.ts:9]: no, we currently don't have any limitations/validations on tags

**Pull request merged**